[UTIL] new RNPxl search engine

timosachsenberg opened the pull request:
Peptide search engine optimized for nucleotide cross-linking experiments.
Added as a util with tests from manual annotations but no extensive documentation at this point.

timosachsenberg commented:
I would suggest to just "squash and merge" to reduce the number of commits.

hroest commented:
unless you believe that the version history has some significant information (e.g. commit comments) we can squash.

hroest commented on src/utils/RNPxlSearch.cpp:
```diff
@@ -0,0 +1,2200 @@
+                  }
+                }
+              } 
+              else
```
some typos here...

hroest replied:
peptides? typo

hroest replied:
peptide -> typo

hroest commented on src/utils/RNPxlSearch.cpp:
```diff
@@ -0,0 +1,2200 @@
+          if (exp_pc_nucleotide_count[pcn_it->first] < pcn_it->second) 
+          {
+            all_present = false;
+            break; // present but to low occurance? -> not applicable
```
typos

hroest commented on src/utils/RNPxlSearch.cpp:
```diff
@@ -0,0 +1,2200 @@
+        if (!all_present) continue;
+
+        // now we need to check if all losses match exactly (ignoring the order) 
+        // TODO: to achieve this we only check if string approximatly match by sorting all characters and then comparing the sorted strings
```
approximatly -> typo

hroest commented on src/utils/RNPxlSearch.cpp:
```diff
@@ -0,0 +1,2200 @@
+              #ifdef DEBUG_RNPXLSEARCH
+              else
+              {
+                LOG_DEBUG << "Charge missmatch in alignment: " << ion_name << " at position: " << fragment_mz << " charge fragment: " << fragment_charge << " theo. charge: " << 1 << endl;
```
mismatch -> typo

hroest commented on src/utils/RNPxlSearch.cpp:
```diff
@@ -0,0 +1,2200 @@
+          for (Size i = 0; i != sites_sum_score.size(); ++i) 
+          {            
+            sites_sum_score[i] = 0.0;
+            // caclculate how many ions are explained by this position
```
typo

hroest commented:
however, my main concern is whether it would be possible to modularize the code in some manner since it currently seems to be a single class with 2000+ lines of code and a 400 LOC main function. Especially for this amount of code, it would also be great to have some more documentation, e.g. describing the general function of the class, what it does and which functions it uses and why...

some functions like spectrum de-isotoping could be at least put in its own namespace and eventually moved into the library as their own classes

other functions like `preprocessSpectra_` should get documentation as the name is very generic and it would be good to document what the function is doing. Similar `postScoreHits_` which quite likely should be broken into multiple functions as it currently is 660+ LOC long.

Other functions like `getPrecursorToFragmentAdducts_` look to me like they could be static functions or even moved into their own class / namespace to signal more clearly when code does not modify the object

also, what happened to xQuest.cpp? seems it was removed at some point, does it still live on somewhere?

hroest commented:
I think an intermediate path before moving stuff into the OpenMS library itself is to create individual, distinct classes that live inside the cpp file and are ready to be moved into the library with proper unit tests as soon as they are used by other code as well (or somebody feels like doing some maintenance work...)

timosachsenberg commented:
I moved most of the scoring to be also used by the now slightly renamed PPXL search engine.
The stuff that is currently inside the tool is still not final in terms that Eugen and me still discuss details on how we can reuse the same functionality. I refactored a bit following your suggestions. postScoreHits is also not final and its structure might change. I can put some more effort in there to break it down but would like to postpone this a bit if possible.

hroest commented:
@timosachsenberg great, it was just a suggestion. However, given how central search engines are in Mass spec and that probably several people would like to try similar things I think in the long run it would be nice to have a shared code base for people writing search engines inside OpenMS. As you say, the question is whether the work is done now or later. Maybe other people can weigh in?

timosachsenberg commented:
I totally agree. @hroest we already have a shared codebase that Eugen's PP cross-link, SimpleSearchEngine and RNPxlSearch uses (either it's already fully integrated into existing classes (via past PRs - like our extension to the digestions, alignment, or the theoretical spectrum generator) or tested but still located at https://github.com/OpenMS/OpenMS/tree/develop/src/openms/include/OpenMS/ANALYSIS/RNPXL )
The remaining parts in our utils are gradually moved to OpenMS when we see that the design converges and we are happy with the performance.
Current discussions cover whether these tools should drop RichPeakSpectrum and only rely on the Integer/Float/StringDataArrays stored with MSSpectrum. We are not fully decided yet but this is another reason why we currently keep stuff in our utils.

hroest commented:
>  Current discussions cover whether these tools should drop RichPeakSpectrum and only rely on the Integer/Float/StringDataArrays stored with MSSpectrum. We are not fully decided yet but this is another reason why we currently keep stuff in our utils.

I would even suggest to go one step further, drop MSSpectrum and only rely on `std::vector<double>` as we have done in some of the OpenSWATH code. It really improved performance and made the interface cleaner and more re-usable. See https://github.com/OpenMS/OpenMS/wiki/OpenMS-2.0-Kernel-Whitepaper#spectrumchromatogram-interface but this is long-term stuff...but it helps to think ahead

timosachsenberg commented:
Would be nice to have but I think neither Eugen nor me has the capacities right now to do this in parallel to the other stuff.

timosachsenberg commented:
any further suggestions?

hroest commented on src/openms/include/OpenMS/KERNEL/MSSpectrum.h:
```diff
@@ -330,6 +330,78 @@ namespace OpenMS
 
+    /// Returns an iterator to the first float data array with the given name. 
+    /// The End iterator is returned in case no data array with given name exists.
+    typename FloatDataArrays::iterator getFloatDataArrayByName(const String& name)
```
I wonder how useful it is to return a start iterator but have no way to access the end iterator? Wouldnt it be easier to return a reference to the whole array and let the user decide how to iterate?

hroest replied:
see below, on second thought I think it is the best solution. Still a bit clumsy as we dont really want to iterate anything here. The other solution would be to return an index and return `-1` in case of failure...that also seems pretty straight forward to me...

hroest commented on src/openms/include/OpenMS/CHEMISTRY/ModificationsDB.h:
```diff
@@ -120,6 +120,16 @@ namespace OpenMS
+    /**
        @brief Returns the index of the modification in the mods_ vector; a unique name must be given
 
        @throw Exception::ElementNotFound if not exactly one matching modification was found
```
can you add a comment that the user should probably call `hasModification` first to avoid an expensive `try - catch` ?

hroest commented:
I am mostly for merge but please change the `getXXXDataArrayByName` functions. I dont like them for the reasons stated above and I strongly opt for a better implementation that works for MSSpectrum and MSChromatograms -- even more so as currently no production code even seems to use these functions.

Also, please fix the typos and run the spellchecker in `tools/spellcheck/`

hroest commented on src/openms/include/OpenMS/KERNEL/MSSpectrum.h:
```diff
@@ -330,6 +330,78 @@ namespace OpenMS
+
+    /// Returns an iterator to the first string data array with the given name. 
+    /// The End iterator is returned in case no data array with given name exists.
+    typename StringDataArrays::iterator getStringDataArrayByName(const String& name)
```
the `end` iterator is not such a bad idea, I think better than anything else. Otherwise we would have to return an empty (static) array but it would be very bad if somebody were to try and insert things in there ...

hroest replied:
PS: you could even template on the type of array passed in and save even more code... ?

timosachsenberg replied:
Ok so the first solution would be to use something like putting this:
```
template <class T>
T::iterator getStringDataArrayByName(const T& name)
{...}
```
in a common header of MSSpectrum / MSChromatogram (+- some additional typename - I usually get those wrong ;)

timosachsenberg replied:
hmm I am pretty bad in getting templated code right on the first try but I think you probably mean something like
```
template <class T>
typename T::StringDataArray::iterator T::getStringDataArrayByName(const String& name)
{...}
in a common header of MSSpectrum / MSChromatogram (+- some additional typename - I usually get those wrong ;)

timosachsenberg commented:
ok changed it to the template version

timosachsenberg commented:
don't merge yet

timosachsenberg commented:
ok also added addModification to pyOpenMS. Good to go now I think...

hroest commented on src/openms/include/OpenMS/KERNEL/SpectrumHelper.h:
```diff
@@ -0,0 +1,63 @@
+  /// Returns an iterator to the first data array with the given name. 
+  /// The end iterator is returned in case no data array with given name exists.
+  template <class DataArrayT>
+  typename DataArrayT::iterator getDataArrayByName(DataArrayT& a, const String& name)
```
I like it much more, seems more elegant to me - and no clutter

timosachsenberg replied:
thanks I added it

timosachsenberg commented:
I made all changes from the additional comments so its good to merge. I decided to keep the commits without squashing btw.
